hist/histv7/inc/ROOT/RHistImpl.hxx

```diff
@@ -227,122 +257,467 @@ namespace Internal {
+template <class... AXISCONFIG>
+int GetNOverflowBinsFromAxes(AXISCONFIG... axisArgs)
+{
+   using axesTuple = std::tuple<AXISCONFIG...>;
+   return RGetNBinsCount<sizeof...(AXISCONFIG) - 1, axesTuple>()(axesTuple{axisArgs...}) - RGetNBinsNoOverCount<sizeof...(AXISCONFIG) - 1, axesTuple>()(axesTuple{axisArgs...});
+}
 
-// Break recursion
-template <class HISTIMPL, class AXES, bool GROW>
-struct RGetBinIndex<-1, HISTIMPL, AXES, GROW> {
-   int operator()(HISTIMPL *, const AXES &, const typename HISTIMPL::CoordArray_t &,
-                  RAxisBase::EFindStatus &status) const
+/// Recursively fills the ranges of all axes, excluding under- and overflow.
+/// Each call fills `range` with `begin()` and `end()` of the current axis, excluding
+/// under- and overflow.
+template <int I, class AXES>
+struct RFillIterRange;
+
+template <class AXES>
+struct RFillIterRange<-1, AXES> {
+   void operator()(Hist::AxisIterRange_t<std::tuple_size<AXES>::value> & /*range*/, const AXES & /*axes*/) const
+   {}
+};
+
+template <int I, class AXES>
+struct RFillIterRange {
+   void operator()(Hist::AxisIterRange_t<std::tuple_size<AXES>::value> &range, const AXES &axes) const
    {
-      status = RAxisBase::EFindStatus::kValid;
-      return 0;
+      range[0][I] = std::get<I>(axes).begin();
+      range[1][I] = std::get<I>(axes).end();
+      RFillIterRange<I - 1, AXES>()(range, axes);
    }
 };
 
-template <int I, class HISTIMPL, class AXES, bool GROW>
-struct RGetBinIndex {
-   int operator()(HISTIMPL *hist, const AXES &axes, const typename HISTIMPL::CoordArray_t &x,
-                  RAxisBase::EFindStatus &status) const
+/// Recursively gets the number of regular bins just before the current dimension.
+/// Each call gets the previous axis' number of regular bins multiplied
+/// by the number of regular bins before the previous axis.
+template <int I, int NDIMS, typename BINS, class AXES>
+struct RGetNRegularBinsBefore;
+
+template <int NDIMS, typename BINS, class AXES>
+struct RGetNRegularBinsBefore<-1, NDIMS, BINS, AXES> {
+   void operator()(BINS &/*binSizes*/, const AXES &/*axes*/) const
+   {}
+};
+
+template <int I, int NDIMS, typename BINS, class AXES>
+struct RGetNRegularBinsBefore {
+   void operator()(BINS &binSizes, const AXES &axes) const
+   {
+      constexpr const int thisAxis = NDIMS - I - 1;
+      binSizes[thisAxis] = binSizes[thisAxis-1] * std::get<thisAxis-1>(axes).GetNBinsNoOver();
+      RGetNRegularBinsBefore<I - 1, NDIMS, BINS, AXES>()(binSizes, axes);
+   }
+};
+
+/// Recursively gets the total number of regular bins before the current dimension.
+/// Each call gets the current axis' number of regular bins before the global_bin
+/// in the current dimension multiplied by the number of regular bins before the
+/// current axis.
+/// If the global_bin is in under- or overflow in the current dimension (local bin),
+/// there is no need to process further.
```

Comment:
First of all, I greatly appreciate the effort you made to document all the templates in the current version of this PR. It helps me greatly at reviewing it, and I think it will also help other people who need to go through the RHist codebase in the future.

I think this particular comment could be clarified a little, by explaining the context a little more:

- We're currently looking at an underflow or overflow bin.
- That bin has its per-axis coordinates spelled out in the `localBins` parameter.
- Those coordinates were also translated to a zero-based which is more convenient for some operations, that's the `virtualBins` parameter.
- We want to know how many regular bins lie before the current overflow bin in the histogram's global binning order (which so far I thought was row-major, but now I'm not sure, maybe it's actually column-major... it doesn't matter, we don't need to spell out what is the global binning order anyway).